### DISCUSSION TITLE
Decay of micrometer’s max values

### DISCUSSION THREAD

**Discussion by lecodeski** [quarkusio/quarkus#28337] [Q&A] [answered]
Hello Quarkus Community 👋 

How Do I control the decay/zeroing/resetting of max values when using Micrometer summaries for Prometheus? I already tried `quarkus.micrometer.export.json.expiry=PT1S` and `quarkus.micrometer.export.json.buffer-length` but that seemed to have no effect.

Thanks for any hint in advance.

[a comment by an automated account is not shown]

**Comment by ebullient** [comment]
json settings will apply to the json registry, not to prometheus.

Which settings are you trying to change? Global settings? For a few metrics in particular?

See the note about the behavior of time window settings here: https://micrometer.io/docs/concepts#_timers

> The maximum statistical value for basic Timer implementations, such as CumulativeTimer and StepTimer, is a time window maximum (TimeWindowMax). It means that its value is the maximum value during a time window. If no new values are recorded for the time window length, the max is reset to 0 as a new time window starts. Time window size is the step size of the meter registry, unless expiry in DistributionStatisticConfig is explicitly set to other value. 
> 
> A time window maximum is used to capture maximum latency in a subsequent interval *after* heavy resource pressure triggers the latency and prevents metrics from being published. Percentiles are also time window percentiles (TimeWindowPercentileHistogram). 

Depending on which you would like (all timers and summaries or specific metrics), you need to create a MeterFilter that returns an updated configuration. There is an example here (note this constrains the distribution summary configuration to "myservice"): https://quarkus.io/guides/micrometer#using-meterfilter-to-configure-metrics

```java
    /** Enable histogram buckets for a specific timer */
    @Produces
    @Singleton
    public MeterFilter enableHistogram() {
        return new MeterFilter() {
            @Override
            public DistributionStatisticConfig configure(Meter.Id id, DistributionStatisticConfig config) {
                if(id.getName().startsWith("myservice")) {
                    return DistributionStatisticConfig.builder()
                        .percentiles(0.5, 0.95)     // median and 95th percentile, not aggregable
                        .percentilesHistogram(true) // histogram buckets (e.g. prometheus histogram_quantile)
                        .build()
                        .merge(config);
                }
                return config;
            }
        };
    }
```

**Reply by lecodeski** [reply]
Whoa, what a fast reply 🤩 and I think you got me on the right track. I think, yeah, I need a MeterFilter. I will try that out later this week (just a bit busy the next days).

Regards you questions:
- The scope of the Filter is not that much relevant, could also be globally
- It should a least catch all max values of all summaries
- Thus a `@MeterFilterConstraint` may be the right choice, but is not mandatory

I will try a MeterFilter and get back here with the results.

Thanks so far, Erin! 🙏

**Reply by lecodeski** [reply, marked as answer]
Okay, couldn’t resist. Took the time to fool around and got it working 🎉 with `expiry()`:

```java
@Singleton
public class CustomConfiguration {
  @Produces
  @Singleton
  public MeterFilter configureAllRegistries() {
    return new MeterFilter() {
      public DistributionStatisticConfig configure(final Meter.Id id, final DistributionStatisticConfig config) {
        return DistributionStatisticConfig.builder()
            .expiry(Duration.ofSeconds(3))
            .build()
            .merge(config);
      }
    };
  }
}
```

So summary:
Though it is not configurable via application.properties, a `MeterFilter` is the key to get this (and a lot of other stuff of micrometer) configured.

Thanks for your hints, @ebullient 👍

**Reply by ebullient** [reply]
Amazing!! Glad to hear you got it working.
